### PR TITLE
Added support for Tuple operations in the CFG.

### PR DESCRIPTION
**Pull request by 333fred**
Tagging @dotnet/roslyn-compiler @dotnet/analyzer-ioperation @AlekseyTs for review.

### CODE REVIEW

**Review by mavasani — DISMISSED** [review]

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs]: >(i1, b ? i2 : i3) = (1, 2); [](start = 8, length = 27)

It would be good to test success scenario like ```(i1, b1 ? ref i2 : ref i3) = (1, 2);```. Also consider making ```i2``` and ```i3``` parameters to avoid "Use of unassigned local variable" diagnostics

**Comment by AlekseyTs** [comment]
>         public override IOperation VisitDeconstructionAssignment(IDeconstructionAssignmentOperation operation, int? captureIdForResult)

Since you are touching deconstruction scenarios in testing, it might make sense to adjust implementation of this function and expand the testing accordingly

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3091 in eacde95. [](commit_id = eacde953c93bb91675c41311cecdf33e1e2e0f62, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs]: >(1, b ? 2 : 3) [](start = 23, length = 14)

It would be good to cover scenarios when control flow is not in the last element.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs]: >} [](start = 8, length = 1)

Consider adding relevant tests for VB, or adding a PROTOTYPE comment in the right file with references to the tests that should be ported from C#.

**Review by AlekseyTs — DISMISSED** [review]
LGTM (iteration 1), with some suggestions

**Comment by 333fred** [comment]
@mavasani @AlekseyTs I dismissed your reviews, as I added IDeconstructionAssignmentOperation support per Aleksey's suggestion. Should be a small additional change for that.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs]: >IFlowCaptureReferenceOperation: 0 [](start = 16, length = 33)

I think tracing deconstruction assignment through a capture reference which is a tuple, which in turn can have a control flow is going to be too complicated and unexpected because language doesn't allow arbitrary expression as a destination. I think we should preserve ITupleOperation as the target of a deconstruction assignment and substitute its elements instead.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs]: >(b ? (i1, i2) : (i3, i4)) = (i5, i6); [](start = 8, length = 37)

I don't think this is a deconstruction

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs]: >IDeclarationExpressionOperation (OperationKind.DeclarationExpression, Type: System.Int32) (Syntax: 'var i1') [](start = 22, length = 108)

This IDeclarationExpressionOperation node doesn't have to be spilled and I think we shouldn't spill it.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs]: >(b ? (var i1, var i2) : (var i3, var i4)) = (1, 2); [](start = 8, length = 51)

I don't think this is a deconstruction

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs]: >IDeclarationExpressionOperation [](start = 16, length = 31)

I don't think we should spill this node

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs]: >IFlowCaptureReferenceOperation [](start = 20, length = 30)

I would expect an IDeclarationExpressionOperation node here

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs]: >IFlowCaptureReferenceOperation [](start = 20, length = 30)

I would expect IDeclarationExpressionOperation of original shape here.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs]: >(var i1, var i2) [](start = 8, length = 16)

Consider also testing a mixed case:
```
int i2;
(int i1, i2) = ...
``` 

This is an error today, but the transformation we do should still make sense.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeclarationExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeclarationExpression.cs]: > M2(b ? out var i1 : out var i2) [](start = 7, length = 32)

Do we even have a declaration expression in the parse tree for this scenario?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3).

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback. I moved all the nondeconstruction test cases out to other files.

---
In reply to: [380953595](https://github.com/dotnet/roslyn/pull/26108#issuecomment-380953595) [](ancestors = 380953595)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >// This is an error case, [](start = 16, length = 25)

I don't think this is accurate, this could be a DeclarationExpression in success scenario.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >PushArray(tupleTarget.Elements); [](start = 16, length = 32)

I don't think this is going to properly handle nested case

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 4)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [381233616](https://github.com/dotnet/roslyn/pull/26108#issuecomment-381233616) [](ancestors = 381233616)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs]: >(i1, (i2, i3)) [](start = 8, length = 14)

Could there be interesting nesting cases with DeclarationExpressions, like ```var (y, (a, b), x??z)```? Basically we don't want capture tuples used within a declaration expressions.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >GetInstance() [](start = 66, length = 13)

Do we know the size of the builder that we need?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs]: This case actually doesn't produce a deconstruction, it's an invalid operation.

---
In reply to: [181502121](https://github.com/dotnet/roslyn/pull/26108#discussion_r181502121) [](ancestors = 181502121)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: We do, I'll update.

---
In reply to: [181503256](https://github.com/dotnet/roslyn/pull/26108#discussion_r181503256) [](ancestors = 181503256)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >return new DeclarationExpression(Visit(operation.Expression), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit); [](start = 12, length = 164)

Consider reusing logic from ```VisitDeconstructionAssignment``` if ```operation.Expression``` is a tuple. I think this will make implementation more robust. We want to guarantee that the shape of the structure is not changed no matter what.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 5), with some suggestions.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: I'll do this in a followup PR, filed [https://github.com/dotnet/roslyn/issues/26151](https://github.com/dotnet/roslyn/issues/26151) to track.

---
In reply to: [181504915](https://github.com/dotnet/roslyn/pull/26108#discussion_r181504915) [](ancestors = 181504915)

**Comment by 333fred** [comment]
Ping @mavasani @dotnet/analyzer-ioperation @dotnet/roslyn-compiler for reviews please.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: I'm surprised by the change here. We don't need to visit `operation`?

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Everything that is on `_evalStack` has already been visited. This extra call to Visit is actually a bug that was causing issues with DeconstructionFlow_07.

---
In reply to: [181832673](https://github.com/dotnet/roslyn/pull/26108#discussion_r181832673) [](ancestors = 181832673)

**Review by jcouv — APPROVED** [review]
LGTM Thanks

**Review by mavasani — APPROVED** [review]

**Comment by 333fred** [comment]
Manually squashing commits before merge.

**Pull request merged**
